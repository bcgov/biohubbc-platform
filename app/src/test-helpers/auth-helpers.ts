import { SYSTEM_ROLE } from 'constants/roles';
import { IAuthState } from 'contexts/authStateContext';
import Keycloak from 'keycloak-js';

export const UnauthenticatedUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: false
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: false,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemAdminAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN],
    hasSystemRole: () => true,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => 'test/login'
  }
};

// Same effect as `Partial` but applies to all levels of a nested object
type Subset<T> = {
  [P in keyof T]?: T[P] extends Record<any, any> | undefined ? Subset<T[P]> : T[P];
};

/**
 * Build and return a mock auth state object.
 *
 * @param {{ base: IAuthState; overrides?: Subset<IAuthState> }} options
 * @return {*}  {IAuthState}
 */
export const getMockAuthState = (options: { base: IAuthState; overrides?: Subset<IAuthState> }): IAuthState => {
  const { base, overrides } = options;

  return {
    ...base,
    ...overrides,
    keycloakWrapper: {
      ...base.keycloakWrapper,
      ...overrides?.keycloakWrapper,
      Keycloak: {
        ...base.keycloakWrapper?.keycloak,
        ...overrides?.keycloakWrapper?.keycloak
      }
    }
  } as unknown as IAuthState;
};
