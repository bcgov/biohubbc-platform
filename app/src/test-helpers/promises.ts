/**
 * 
 */
export class Deferred<R = void, Q = void> {
    resolve: (value: R) => void = () => {};
    reject: (reason?: Q) => void = () => {};
    promise: Promise<R>;

    constructor() {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }

    recycle = () => {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }
}
