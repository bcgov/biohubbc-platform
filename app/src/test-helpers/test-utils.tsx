import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import appTheme from 'themes/appTheme'

const AllProviders: React.FC<React.PropsWithChildren> = (props) => {
  const { children} = props
  return (
    <ThemeProvider theme={appTheme}>    
      {children}
    </ThemeProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
