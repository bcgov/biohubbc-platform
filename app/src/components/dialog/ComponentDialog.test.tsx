import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import ComponentDialog from 'components/dialog/ComponentDialog';
import { fireEvent, render, waitFor } from 'test-helpers/test-utils';

const handleOnClose = jest.fn();

const TestComponent = () => {
  return (
    <Box>
      <Typography>This is a test component</Typography>
    </Box>
  );
};

const renderContainer = ({ dialogTitle, open = true }: { dialogTitle: string; open?: boolean }) => {
  return render(
    <div id="root">
      <ComponentDialog dialogTitle={dialogTitle} open={open} onClose={handleOnClose}>
        <TestComponent />
      </ComponentDialog>
    </div>
  );
};

describe('ComponentDialog', () => {
  it('Does not render dialog pop up when open set to false', () => {
    const { queryByText } = renderContainer({ dialogTitle: 'this is a test', open: false });

    expect(queryByText('this is a test')).toEqual(null);
  });

  it('Does render dialog pop up when open set to true', () => {
    const { getByText } = renderContainer({ dialogTitle: 'this is a test', open: true });

    expect(getByText('this is a test')).toBeVisible();
  });

  it('calls the onClose prop when `Close` button is clicked', async () => {
    const { findByText } = renderContainer({ dialogTitle: 'this is a test' });

    const CloseButton = await findByText('Close', { exact: false });

    fireEvent.click(CloseButton);

    await waitFor(() => {
      expect(handleOnClose).toHaveBeenCalledTimes(1);
    });
  });
});
