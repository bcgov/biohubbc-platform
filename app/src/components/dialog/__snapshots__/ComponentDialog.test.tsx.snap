// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentDialog matches snapshot when open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root"
    />
  </div>
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby="component-dialog-description"
        aria-labelledby="component-dialog-title"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXl css-2rbg70-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-omylgo-MuiTypography-root-MuiDialogTitle-root"
          id="component-dialog-title"
        >
          this is a test
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-12btgb4-MuiTypography-root"
            >
              This is a test component
            </p>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-1qm3mg4-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1eys6cy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`ComponentDialog matches the snapshot when not open 1`] = `
<body>
  <div>
    <div
      id="root"
    />
  </div>
</body>
`;
